README
NAME
    Math::Random::OO - Consistent object-oriented interface for generating
    random numbers

SYNOPSIS
      # Using factory functions
      use Math::Random::OO qw( Uniform UniformInt );
      push @prngs, Uniform(), UniformInt(1,6);
     
  # Explicit creation of subclasses
      use Math::Random::OO::Normal;
      push @prngs, Math::Random::OO::Normal->new(0,2);
     
  $_->seed(23) for (@prngs);
      print( $_->next(), "\n") for (@prngs);

DESCRIPTION
    CPAN contains many modules for generating random numbers in various ways
    and from various probability distributions using pseudo-random number
    generation algorithms or other entropy sources. (The "SEE ALSO" section
    has some examples.) Unfortunately, no standard interface exists across
    these modules. This module defines an abstract interface for random
    number generation. Subclasses of this model will implement specific
    types of random number generators or will wrap existing random number
    generators.

    This consistency will come at the cost of some efficiency, but will
    enable generic routines to be written that can manipulate any provided
    random number generator that adheres to the interface. E.g., a
    stochastic simulation could take a number of user-supplied parameters,
    each of which is a Math::Random::OO subclass object and which represent
    a stochastic variable with a particular probability distribution.

USAGE
  Factory Functions
     use Math::Random::OO qw( Uniform UniformInt Normal Bootstrap );
     $uniform = Uniform(-1,1);
     $uni_int = UniformInt(1,6);
     $normal  = Normal(1,1);
     $boot    = Bootstrap( 2, 3, 3, 4, 4, 4, 5, 5, 5 );

    In addition to defining the abstract interface for subclasses, this
    module imports subclasses and exports factory functions upon request to
    simplify creating many random number generators at once without typing
    "Math::Random::OO::Subclass->new()" each time. The factory function
    names are the same as the suffix of the subclass following
    "Math::Random::OO". When called, they pass their arguments directly to
    the "new" constructor method of the corresponding subclass and return a
    new object of the subclass type. Supported functions and their
    subclasses include:

    *   "Uniform" -- Math::Random::OO::Uniform (uniform distribution over a
        range)

    *   "UniformInt" -- Math::Random::OO::UniformInt (uniform distribution
        of integers over a range)

    *   "Normal" -- Math::Random::OO::Normal (normal distribution with
        specified mean and standard deviation)

    *   "Bootstrap" -- Math::Random::OO::Bootstrap (bootstrap resampling
        from a non-parameteric distribution)

INTERFACE
    All Math::Random::OO subclasses must follow a standard interface. They
    must provide a "new" method, a "seed" method, and a "next" method.
    Specific details are left to each interface.

  "new"
    This is the standard constructor. Each subclass will define parameters
    specific to the subclass.

  "seed"
     $prng->seed( @seeds );

    This method takes seed (or list of seeds) and uses it to set the initial
    state of the random number generator. As some subclasses may optionally
    use/require a list of seeds, the interface mandates that a list must be
    acceptable. Generators requiring a single seed must use the first value
    in the list.

    As seeds may be passed to the built-in "srand()" function, they may be
    truncated as integers, so 0.12 and 0.34 would be the same seed.

  "next"
     $rnd = $prng->next();

    This method returns the next random number from the random number
    generator. It does not take (and must not use) any parameters.

BUGS
    Please report bugs using the CPAN Request Tracker at
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Math-Random-OO

AUTHOR
    David A Golden <dagolden@cpan.org>

    http://dagolden.com/

COPYRIGHT
    Copyright (c) 2004, 2005 by David A. Golden

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

SEE ALSO
    This is not an exhaustive list -- search CPAN for that -- but represents
    some of the more common or established random number generators that
    I've come across.

    Math::Random -- multiple random number generators for different
    distributions (a port of the C randlib)
    Math::Rand48 -- perl bindings for the drand48 library (according to
    perl56delta, this may already be the default after perl 5.005_52 if
    available)
    Math::Random::MT -- The Mersenne Twister PRNG (good and fast)
    Math::TrulyRandom -- an interface to random numbers from interrupt
    timing discrepancies



Changes
Revision history for Perl module Math::Random::OO

{{$NEXT}}

0.22      2013-06-09 15:07:27 America/New_York

    - Removed useless voodoo math [RT#85314]
    - Fixed test bug with negative seeds [RT#86002]
    - Converted to Dist::Zilla
    - Changed license to Apache 2
    - Changed bugtracker

0.21 Sat May  2 07:40:03 EDT 2009
    - Fixed major seed() bug (RT#45630)

0.20 Fri Jul 21 08:00:34 EDT 2006
    - Cleaned up pod coverage test

0.19 Fri Jul 21 07:00:34 EDT 2006
    - Whee!  Added back pod/coverage tests in a sane way.

0.18 Fri Dec 16 13:20:26 EST 2005
    - really remove Test::Exception as a dependency

0.17 Thu Dec 15 01:39:14 EST 2005
    - added missing Class::Accessor::Fast dependency to Build.PL
    - removed Test::Exception and Attribute::Util dependencies from
      tests and code

0.16 Mon Aug  1 11:10:28 EDT 2005
    - build_requires moved to requires to work around CPANPLUS bug

0.15 Thu Jun  9 15:59:11 EDT 2005
    - removed pod/podcoverage/distribution tests to minimize dependencies

0.14 Tue Jun  7 18:10:09 EDT 2005
    - fixed dependencies in Build.PL
    - changed Makefile compatibility

0.13 Wed Mar  9 06:30:16 EST 2005
    - added pod and pod coverage testing

0.12 Wed Feb  2 22:50:30 EST 2005
    - fixed test suite for new version of Test::MockRandom
    - improved test coverage

0.11 Tue Jan  4 07:20:38 EST 2005
    - Fixed POD for better hyperlinking on CPAN
    - Refactored tests
    - Added Test::Distribution to automated tests (but not required
      for users)
    - Replaced Test::Float with Test::Number::Delta

0.10 Tue Dec 28 12:22:04 EST 2004
    - initial public release

0.01 Tue Dec 14 13:36:52 2004
	- original version; created by ExtUtils::ModuleMaker::TT


lib/Math/Random/OO.pm
use 5.006;
use strict;
use warnings;

package Math::Random::OO;
# ABSTRACT: Consistent object-oriented interface for generating random numbers
# VERSION

use Carp;

sub import {
    my ( $class, @symbols ) = @_;
    my $caller = caller;
    for (@symbols) {
        no strict 'refs';
        my $subclass = "Math::Random::OO::$_";
        eval "require $subclass";
        *{"${caller}::$_"} = sub { return ${subclass}->new(@_) };
    }
}

sub new {
    my $class = shift;
    return bless( {}, ref($class) ? ref($class) : $class );
}

sub seed { die "call to abstract method 'seed'" }

sub next { die "call to abstract method 'next'" }

1;

__END__

=head1 SYNOPSIS

  # Using factory functions
  use Math::Random::OO qw( Uniform UniformInt );
  push @prngs, Uniform(), UniformInt(1,6);
 
  # Explicit creation of subclasses
  use Math::Random::OO::Normal;
  push @prngs, Math::Random::OO::Normal->new(0,2);
 
  $_->seed(23) for (@prngs);
  print( $_->next(), "\n") for (@prngs);

=head1 DESCRIPTION

CPAN contains many modules for generating random numbers in various ways
and from various probability distributions using pseudo-random number
generation algorithms or other entropy sources.  (The L</"SEE ALSO"> section
has some examples.)  Unfortunately, no standard interface exists across these
modules.  This module defines an abstract interface for random number
generation.  Subclasses of this model will implement specific types of random
number generators or will wrap existing random number generators.

This consistency will come at the cost of some efficiency, but will enable
generic routines to be written that can manipulate any provided random number
generator that adheres to the interface.  E.g., a stochastic simulation could
take a number of user-supplied parameters, each of which is a Math::Random::OO
subclass object and which represent a stochastic variable with a particular
probability distribution.

=head1 USAGE

=head2 Factory Functions

 use Math::Random::OO qw( Uniform UniformInt Normal Bootstrap );
 $uniform = Uniform(-1,1);
 $uni_int = UniformInt(1,6);
 $normal  = Normal(1,1);
 $boot    = Bootstrap( 2, 3, 3, 4, 4, 4, 5, 5, 5 );
 
In addition to defining the abstract interface for subclasses, this module
imports subclasses and exports factory functions upon request to simplify
creating many random number generators at once without typing
C<Math::Random::OO::Subclass-E<gt>new()> each time.  The factory function names are
the same as the suffix of the subclass following C<Math::Random::OO>.  When
called, they pass their arguments directly to the C<new> constructor method of
the corresponding subclass and return a new object of the subclass type.
Supported functions and their subclasses include:

=over

=item *

C<Uniform> -- L<Math::Random::OO::Uniform> (uniform distribution over a range)

=item *

C<UniformInt> -- L<Math::Random::OO::UniformInt> (uniform distribution of
integers over a range)

=item *

C<Normal> -- L<Math::Random::OO::Normal> (normal distribution with specified
mean and standard deviation)

=item *

C<Bootstrap> -- L<Math::Random::OO::Bootstrap> (bootstrap resampling from
a non-parameteric distribution)

=back

=head1 INTERFACE

All Math::Random::OO subclasses must follow a standard interface.  They must
provide a C<new> method, a C<seed> method, and a C<next> method.  Specific 
details are left to each interface.

=head2 C<new>

This is the standard constructor.  Each subclass will define parameters specific to the subclass.

=head2 C<seed>

 $prng->seed( @seeds );

This method takes seed (or list of seeds) and uses it to set the initial state
of the random number generator.  As some subclasses may optionally use/require
a list of seeds, the interface mandates that a list must be acceptable.
Generators requiring a single seed must use the first value in the list.

As seeds may be passed to the built-in C<srand()> function, they may be 
truncated as integers, so 0.12 and 0.34 would be the same seed.  Only
positive integers should be used.

=head2 C<next>

 $rnd = $prng->next();

This method returns the next random number from the random number generator.
It does not take (and must not use) any parameters. 

=head1 SEE ALSO

This is not an exhaustive list -- search CPAN for that -- but represents some of
the more common or established random number generators that I've come across.

=over

=item L<Math::Random> -- multiple random number generators for different
distributions (a port of the C randlib)

=item L<Math::Rand48> -- perl bindings for the drand48 library (according to
perl56delta, this may already be the default after perl 5.005_52 if available)

=item L<Math::Random::MT> -- The Mersenne Twister PRNG (good and fast)

=item L<Math::TrulyRandom> -- an interface to random numbers from interrupt timing discrepancies

=back

=cut


lib/Math/Random/OO/Uniform.pm
use 5.006;
use strict;
use warnings;

package Math::Random::OO::Uniform;
# ABSTRACT: Generates random numbers from the uniform distribution
# VERSION

# Required modules
use Carp;
use Params::Validate ':all';

# ISA
use base qw( Class::Accessor::Fast );

=method C<new>

 $prng1 = Math::Random::OO::Uniform->new();
 $prng2 = Math::Random::OO::Uniform->new($high);
 $prng3 = Math::Random::OO::Uniform->new($low,$high);

C<new> takes up to two optional parameters and returns a new
C<Math::Random::OO::Uniform> object.  With no parameters, the object generates
random numbers in the range of zero (inclusive) to one (exclusive).  With a
single parameter, the object generates random numbers from zero (inclusive) to
the value of the parameter (exclusive).  Note, the object does this with
multiplication, so if the parameter is negative, the function will return
negative numbers.  This is a feature or bug, depending on your point of view.
With two parameters, the object generates random numbers from the first
parameter (inclusive) to the second parameter (exclusive).  (Actually, as 
long as you have two parameters, C<new> will put them in the right order).

=cut

{
    my $param_spec = {
        low  => { type => SCALAR },
        high => { type => SCALAR }
    };

    __PACKAGE__->mk_accessors( keys %$param_spec );
    #__PACKAGE__->mk_ro_accessors( keys %$param_spec );

    sub new {
        my $class = shift;
        my $self = bless {}, ref($class) ? ref($class) : $class;
        if ( @_ > 1 ) {
            my ( $low, $high ) = sort { $a <=> $b } @_[ 0, 1 ]; # DWIM
            $self->low($low);
            $self->high($high);
        }
        elsif ( @_ == 1 ) {
            $self->low(0);
            $self->high( $_[0] );
        }
        else {
            $self->low(0);
            $self->high(0);
        }
        return $self;
    }
}

=method C<seed>

 $rv = $prng->seed( @seeds );

This method seeds the random number generator.  At the moment, only the
first seed value matters.  It should be a positive integer.

=cut

sub seed {
    my $self = shift;
    srand( $_[0] );
}

=method C<next>

 $rnd = $prng->next();

This method returns the next random number from the random number generator.
It does not take any parameters.

=cut

sub next {
    my ($self) = @_;
    my $rnd = rand( $self->high - $self->low ) + $self->low;
    return $rnd;
}

1;

__END__

=head1 SYNOPSIS

  use Math::Random::OO::Uniform;
  push @prngs,
      Math::Random::OO::Uniform->new(),     # range [0,1)
      Math::Random::OO::Uniform->new(5),    # range [0,5)
      Math::Random::OO::Uniform->new(-1,1); # range [-1,1)
  $_->seed(42) for @prngs;
  print( $_->next() . "\n" ) for @prngs;
  
=head1 DESCRIPTION

This subclass of L<Math::Random::OO> generates random reals from a uniform
probability distribution.

=cut


lib/Math/Random/OO/Bootstrap.pm
use 5.006;
use strict;
use warnings;

package Math::Random::OO::Bootstrap;
# ABSTRACT: Generate random numbers with bootstrap resampling from a non-parametric distribution
# VERSION

# Required modules
use Carp;
use Params::Validate 0.76 ':all';

# ISA
use base qw( Class::Accessor::Fast );

=method C<new>

 $prng = Math::Random::OO::Bootstrap->new(@sample);
 $prng = Math::Random::OO::Bootstrap->new(\@sample);

C<new> takes either a list or a reference to an array containing a
set of observations and returns a new C<Math::Random::OO::Bootstrap> object.
If a reference is provided, the object will make an internal copy
of the array to avoid unexpected results if the source reference
is modified. 

If the desired sample is an array of array references, the list
must be enclosed in an anonymous array reference to avoid ambiguity.

 @sample = ( [ 1, 2, 3], [2, 3, 4] );

 # Correct way
 $prng = Math::Random::OO::Bootstrap->new( [ @sample ] );
 
 # Incorrect -- will only use [1, 2, 3] as the desired sample
 $prng = Math::Random::OO::Bootstrap->new( @sample );

It is an error to call C<new> with no arguments.

=cut

{
    my $param_spec = {
        data => { type => ARRAYREF },
        size => { type => SCALAR }
    };

    __PACKAGE__->mk_accessors( keys %$param_spec );
    #__PACKAGE__->mk_ro_accessors( keys %$param_spec );

    sub new {
        my $class = shift;
        my $self = bless {}, ref($class) ? ref($class) : $class;
        if ( @_ == 0 ) {
            croak 'Math::Random::OO::Bootstrap->new() requires an argument';
        }
        $self->data( ref $_[0] eq 'ARRAY' ? [ @{ $_[0] } ] : [@_] );
        $self->size( scalar @{ $self->data } );
        return $self;
    }
}

=method C<seed>

 $rv = $prng->seed( @seeds );

This method seeds the random number generator.  At the moment, only the
first seed value matters.  It should be a positive integer.

=cut

sub seed {
    my $self = shift;
    srand( $_[0] );
}

=method C<next>

 $rnd = $prng->next();

This method returns the next random number from the random number generator
by resampling with replacement from the provided data. It does not take any
parameters.

=cut

sub next {
    my ($self) = @_;
    my $rnd = int( rand( $self->size ) ); # index 0 to (size-1)
    return $self->data->[$rnd];
}

1;

__END__

=head1 SYNOPSIS

  use Math::Random::OO::Bootstrap;
  @sample = qw( 2 3 3 4 4 5 5 6 6 7 );
  $prng = Math::Random::OO::Bootstrap->new(@sample);
  $prng->seed(42);
  $prng->next() # draws randomly from the sample
  
=head1 DESCRIPTION

This subclass of L<Math::Random::OO> generates random numbers with bootstrap
resampling (i.e. resampling with replacement) from a given set of observations.
Each item in the sample array is drawn with equal probability.

=cut



lib/Math/Random/OO/UniformInt.pm
use 5.006;
use strict;
use warnings;

package Math::Random::OO::UniformInt;
# ABSTRACT: Generates random integers with uniform probability
# VERSION

# Required modules
use Carp;
use Params::Validate ':all';

# ISA
use base qw( Class::Accessor::Fast );

=method C<new>

 $prng1 = Math::Random::OO::UniformInt->new();
 $prng2 = Math::Random::OO::UniformInt->new($high);
 $prng3 = Math::Random::OO::UniformInt->new($low,$high);

C<new> takes up to two optional parameters and returns a new
C<Math::Random::OO::UniformInt> object.  Unlike Uniform, it returns integers
inclusive of specified endpoints. With no parameters, the object generates
random integers in the range of zero (inclusive) to one (inclusive).  With a
single parameter, the object generates random numbers from zero (inclusive) to
the value of the parameter (inclusive).  Note, the object does this with
multiplication, so if the parameter is negative, the function will return
negative numbers.  This is a feature or bug, depending on your point of view.
With two parameters, the object generates random integers from the first
parameter (inclusive) to the second parameter (inclusive).  (Actually, as long
as you have two parameters, C<new> will put them in the right order).  If
parameters are non-integers, they will be truncated to integers before the
range is calculated.  I.e., C<new(-1.2, 3.6)> is equivalent to C<new(-1,3)>.

=cut

{
    my $param_spec = {
        low  => { type => SCALAR },
        high => { type => SCALAR }
    };

    __PACKAGE__->mk_accessors( keys %$param_spec );
    #__PACKAGE__->mk_ro_accessors( keys %$param_spec );

    sub new {
        my $class = shift;
        my $self = bless {}, ref($class) ? ref($class) : $class;
        if ( @_ > 1 ) {
            my ( $low, $high ) = sort { $a <=> $b } @_[ 0, 1 ]; # DWIM
            $self->low( int($low) );
            $self->high( int($high) );
        }
        elsif ( @_ == 1 ) {
            $self->low(0);
            $self->high( int( $_[0] ) );
        }
        else {
            $self->low(0);
            $self->high(1);
        }
        return $self;
    }
}

=method C<seed>

 $rv = $prng->seed( @seeds );

This method seeds the random number generator.  At the moment, only the
first seed value matters.  It should be a positive integer.

=cut

sub seed {
    my $self = shift;
    srand( $_[0] );
}

=method C<next>

 $rnd = $prng->next();

This method returns the next random number from the random number generator.
It does not take any parameters.

=cut

sub next {
    my ($self) = @_;
    my $rnd = int( rand( $self->high - $self->low + 1 ) ) + $self->low;
    return $rnd;
}

1;

__END__

=head1 SYNOPSIS

  use Math::Random::OO::UniformInt;
  push @prngs,
      Math::Random::OO::UniformInt->new(),     # 0 or 1
      Math::Random::OO::UniformInt->new(5),    # 0, 1, 2, 3, 4, or 5
      Math::Random::OO::UniformInt->new(-1,1); # -1, 0, or 1
  $_->seed(42) for @prngs;
  print( $_->next() . "\n" ) for @prngs;
  
=head1 DESCRIPTION

This subclass of L<Math::Random::OO> generates random integers with uniform
probability.

=cut


lib/Math/Random/OO/Normal.pm
use 5.006;
use strict;
use warnings;

package Math::Random::OO::Normal;
# ABSTRACT: Generates random numbers from the normal (Gaussian) distribution
# VERSION

# Required modules
use Carp;
use Params::Validate ':all';

# ISA
use base qw( Class::Accessor::Fast );

=method C<new>

 $prng1 = Math::Random::OO::Normal->new();
 $prng2 = Math::Random::OO::Normal->new($mean);
 $prng3 = Math::Random::OO::Normal->new($mean,$stdev);

C<new> takes up to two optional parameters and returns a new
C<Math::Random::OO::Normal> object.  With no parameters, the object generates
random numbers from the standard normal distribution (mean zero, standard
deviation one).  With a single parameter, the object generates random numbers
with mean equal to the parameter and standard deviation of one.  With two
parameters, the object generates random numbers with mean equal to the first
parameter and standard deviation equal to the second parameter.  (Standard
deviation should be positive, but this module takes the absolute value of the
parameter just in case.)

=cut

{
    my $param_spec = {
        mean  => { type => SCALAR },
        stdev => { type => SCALAR }
    };

    __PACKAGE__->mk_accessors( keys %$param_spec );
    #__PACKAGE__->mk_ro_accessors( keys %$param_spec );

    sub new {
        my $class = shift;
        my $self = bless {}, ref($class) ? ref($class) : $class;
        if ( @_ > 1 ) {
            $self->mean( $_[0] );
            $self->stdev( abs( $_[1] ) );
        }
        elsif ( @_ == 1 ) {
            $self->mean( $_[0] );
            $self->stdev(1);
        }
        else {
            $self->mean(0);
            $self->stdev(1);
        }
        return $self;
    }
}

=method C<seed>

 $rv = $prng->seed( @seeds );

This method seeds the random number generator.  At the moment, only the
first seed value matters.  It should be a positive integer.

=cut

sub seed {
    my $self = shift;
    srand( $_[0] );
}

=method C<next>


 $rnd = $prng->next();

This method returns the next random number from the random number generator.
It does not take any parameters.

=cut

sub next {
    my ($self) = @_;
    my $rnd = rand() || 1e-254; # can't have zero for normals
    return _ltqnorm($rnd) * $self->stdev + $self->mean;
}

#--------------------------------------------------------------------------#
# Function for inverse cumulative normal
# Used with permission
# http://home.online.no/~pjacklam/notes/invnorm/impl/acklam/perl/
#
# Input checking removed by DAGOLDEN as the input will be prechecked
#--------------------------------------------------------------------------#

#<<< No perltidy
sub _ltqnorm {
    # Lower tail quantile for standard normal distribution function.
    #
    # This function returns an approximation of the inverse cumulative
    # standard normal distribution function.  I.e., given P, it returns
    # an approximation to the X satisfying P = Pr{Z <= X} where Z is a
    # random variable from the standard normal distribution.
    #
    # The algorithm uses a minimax approximation by rational functions
    # and the result has a relative error whose absolute value is less
    # than 1.15e-9.
    #
    # Author:      Peter J. Acklam
    # Time-stamp:  2000-07-19 18:26:14
    # E-mail:      pjacklam@online.no
    # WWW URL:     http://home.online.no/~pjacklam

    my $p = shift;
    # DAGOLDEN: arg checking will be done earlier
#    die "input argument must be in (0,1)\n" unless 0 < $p && $p < 1;

    # Coefficients in rational approximations.
    my @a = (-3.969683028665376e+01,  2.209460984245205e+02,
             -2.759285104469687e+02,  1.383577518672690e+02,
             -3.066479806614716e+01,  2.506628277459239e+00);
    my @b = (-5.447609879822406e+01,  1.615858368580409e+02,
             -1.556989798598866e+02,  6.680131188771972e+01,
             -1.328068155288572e+01 );
    my @c = (-7.784894002430293e-03, -3.223964580411365e-01,
             -2.400758277161838e+00, -2.549732539343734e+00,
              4.374664141464968e+00,  2.938163982698783e+00);
    my @d = ( 7.784695709041462e-03,  3.224671290700398e-01,
              2.445134137142996e+00,  3.754408661907416e+00);

    # Define break-points.
    my $plow  = 0.02425;
    my $phigh = 1 - $plow;

    # Rational approximation for lower region:
    if ( $p < $plow ) {
       my $q  = sqrt(-2*log($p));
       return ((((($c[0]*$q+$c[1])*$q+$c[2])*$q+$c[3])*$q+$c[4])*$q+$c[5]) /
               (((($d[0]*$q+$d[1])*$q+$d[2])*$q+$d[3])*$q+1);
    }

    # Rational approximation for upper region:
    if ( $phigh < $p ) {
       my $q  = sqrt(-2*log(1-$p));
       return -((((($c[0]*$q+$c[1])*$q+$c[2])*$q+$c[3])*$q+$c[4])*$q+$c[5]) /
                (((($d[0]*$q+$d[1])*$q+$d[2])*$q+$d[3])*$q+1);
    }

    # Rational approximation for central region:
    my $q = $p - 0.5;
    my $r = $q*$q;
    return ((((($a[0]*$r+$a[1])*$r+$a[2])*$r+$a[3])*$r+$a[4])*$r+$a[5])*$q /
           ((((($b[0]*$r+$b[1])*$r+$b[2])*$r+$b[3])*$r+$b[4])*$r+1);
}
#>>>

1;

__END__

=head1 SYNOPSIS

  use Math::Random::OO::Normal;
  push @prngs,
      Math::Random::OO::Normal->new(),     # mean 0, stdev 1
      Math::Random::OO::Normal->new(5),    # mean 5, stdev 1
      Math::Random::OO::Normal->new(1,3);  # mean 1, stdev 3
  $_->seed(42) for @prngs;
  print( $_->next() . "\n" ) for @prngs;

=head1 DESCRIPTION

This subclass of L<Math::Random::OO> generates random reals from the normal 
probability distribution, also called the Gaussian or bell-curve distribution.

The module generates random normals from the inverse of the cumulative 
normal distribution using an approximation algorithm developed by Peter J. 
Acklam and released into the public domain.  This algorithm claims a
relative error of less than 1.15e-9 over the entire region.

See http://home.online.no/~pjacklam/notes/invnorm/ for details and discussion.

=cut
